### ISSUE TITLE
Nodes removed/reinserted when they are simply shifted

### ISSUE THREAD

**Issue by yjbanov** [angular/angular#3442]
There are reports that `*ng-for` removes and then reinserts nodes when their indices shift. We should make sure that we are not doing that.

**Comment by skybrian** [comment]
Here is my (perhaps unrealistic) test code. I don't know that any real apps suffer from this.

import 'dart:html';

import 'package:angular2/angular2.dart';
import 'package:angular2/src/reflection/reflection.dart' show reflector;
import 'package:angular2/src/reflection/reflection_capabilities.dart'
show ReflectionCapabilities;

void main() {
  reflector.reflectionCapabilities = new ReflectionCapabilities();
  bootstrap(Grid);
}

@Component(
    selector: 'grid'
)
@View(
    template: '''
<div>
  <button (click)="play(up)">Up</button>
  <button (click)="play(down)">Down</button>
  <button (click)="play(left)">Left</button>
  <button (click)="play(right)">Right</button>
  <button (click)="play(stop)">Stop</button>
</div>
<table>
  <tr *ng-for="#row of rows">
    <td *ng-for="#cell of row">{{ cell }}</td>
  </tr>
</table>
''', directives: const [NgFor]
)
class Grid {
  List rows = makeGrid(100, 20);
  Function step;
  bool willRender = false;

  Grid() {
    play(left);
  }

  void play(Function newStep) {
    step = newStep;
    requestRender();
  }

  void requestRender() {
    if (!willRender) {
      window.animationFrame.then((_) => render());
      willRender = true;
    }
  }

  void render() {
    willRender = false;
    if (step != null) {
      step();
      requestRender();
    }
  }

  void up() => rotateUp(rows);
  void down() => rotateDown(rows);
  void left() => rotateLeft(rows);
  void right() => rotateRight(rows);
  final Function stop = null;
}

List<List<int>> makeGrid(int width, int height) {
  var rows = [];
  for (int y = 0; y < height; y++) {
    var row = [];
    for (int x = 0; x < width; x++) {
      row.add(((x~/10)+(y~/10)) % 10);
    }
    rows.add(row);
  }
  return rows;
}

void rotateUp(List<List<int>> rows) {
  var row = rows.removeAt(0);
  rows.add(row);
}

void rotateDown(List<List<int>> rows) {
  var row = rows.removeLast();
  rows.insert(0, row);
}

void rotateLeft(List<List<int>> rows) {
  for (var row in rows) {
    var item = row.removeAt(0);
    row.add(item);
  }
}

void rotateRight(List<List<int>> rows) {
  for (var row in rows) {
    var item = row.removeLast();
    row.insert(0, item);
  }
}

**Comment by mhevery** [comment]
This may be intentional. Could we have a plunker which demonstrates it? How do you determine that something is reinserted?

**Comment by skybrian** [comment]
I don't know what a plunker is. I profiled the test code (above) in Chrome Dev Tools and found that most of the time spent was in functions in NgFor. Internal discussion:
https://groups.google.com/a/google.com/forum/#!topic/dart-angular/_YYS6liombQ

A good heuristic is to find common prefixes and suffixes starting from both the beginning and ends of the lists, and then run a more sophisticated algorithm on whatever is left in the middle. In the common case where there is only one change, this is optimal.

I looked at the diff algorithm a while ago, and it looked like it would handle common prefixes well but not common suffixes.

More ideas about cheap diff heuristics here:
https://neil.fraser.name/writing/diff/

**Comment by yjbanov** [comment]
Longest increasing subsequence (LIS) would be able to identify minimal moves in N*log(N). It's not too hard to implement, but we also need to find the shortest list length past which LIS is faster than a simpler approach. Certainly for lists of 0-2 we should just use ad hoc functions instead of generic list differ.

@mhevery why do you need plunker? @skybrian provided a repro sample in a github gist.

**Comment by vicb** [comment]
The current algorithm outputs added, removed & moved elements (not sure about how it would handle common suffix vs common prefix).

It's probably more how the directive deals with changes that is to blame here ?
1 - remove the view for removed & moved nodes (removes DOM nodes),
2 - add views for inserted nodes.

At this point we have the final number of views but their context might be wrong.

3 - Loop on all the views to set their context (that the index + the element).

First step would be to investigate if the diff algo or the directive code is to blame.

/cc @rkirov

**Comment by bertrandg** [comment]
I have a plunker showing this behavior (with alpha 44) here:
http://plnkr.co/edit/HLbOaD?p=preview

1. Remove the first element
2. Add a new element
3. See that order in the view is not right and not identical as inside the console log
4. Click 2 times on "SHOW/HIDE" button
5. See that order in the view is now correct

Like someone said on stackoverflow, this occurs only with syntax:
```html
<template ng-for #m [ng-for-of]="group.children" #i="index">...</template>
```
Not with this syntax:
```html
<div *ng-for="#m of group.children; #i=index">...</div>
```

([posted on stackoverflow first](http://stackoverflow.com/questions/33363987/angular2-ngfor-inside-tree-model-wrong-order-when-remove-and-then-add-elements))

**Comment by yjbanov** [comment]
Bumping priority on this. I saw this briefly while debugging an app. The wrong insertion order was triggered by non-idempotent expression bound to `ng-for`. Making it idempotent fixed the issue, however, it shouldn't have. Even with non-idempotent expressions we should still render the list correctly.

/cc @naomiblack for coordination

**Comment by rkirov** [comment]
There are two things conflated in this issue:
- when objects in the ng-for array are moved, what should be the corresponding mutations to the DOM. This is a question of efficiency. #1377 captures this issue it a lot better, so I suggest we continue the discussion there.

- reports for actual bugs in the ng-for implementation by @bertrandg but the repro case is quite involved. @bertrandg can you simplify it a bit, I think the minimal repro involves an ng-for wrapping ng-if. It is very off that with the *ng-for syntax the bug disappears.

btw, I ported @skybrian's demo to plunker -http://plnkr.co/edit/h5Q5EtxB0aqNnGVg6AVo?p=preview for future reference.

**Comment by tbosch** [comment]
The bug of @bertrandg is a dupe of #6304 and is solved in #6878.
And as the performance problems is tracked in #1377 I am closing this issue.
